paperqa/clients/crossref.py

```diff
@@ -0,0 +1,331 @@
+    # get DOI via crossref
+    url = f"https://api.crossref.org/works/{quote(doi, safe='')}/transform/application/x-bibtex"
+    async with session.get(url, headers=crossref_headers()) as r:
+        if not r.ok:
+            raise DOINotFoundError(
+                f"Per HTTP status code {r.status}, could not resolve DOI {doi}."
+            )
+        data = await r.text()
+    # must make new key
+    key = data.split("{")[1].split(",")[0]
+    new_key = remove_substrings(key, FORBIDDEN_KEY_CHARACTERS)
+    substrings_to_remove_per_field = {"author": [" and ", ","]}
+    fragments = [
+        remove_substrings(
+            bibtex_field_extract(
+                data, field, missing_replacements=missing_replacements
+            ),
+            substrings_to_remove_per_field.get(field, []),
+        )
+        for field in ("author", "year", "title")
+    ]
+    # replace the key if all the fragments are present
+    if all(fragments):
+        new_key = remove_substrings(("".join(fragments)), FORBIDDEN_KEY_CHARACTERS)
+    # we use the count parameter below to ensure only the 1st entry is replaced
+    return data.replace(key, new_key, 1)
+
+
+async def parse_crossref_to_doc_details(  # noqa: C901
+    message: dict[str, Any],
+    session: aiohttp.ClientSession,
+    query_bibtex: bool = True,
+) -> DocDetails:
+
+    bibtex_source = "self_generated"
+    bibtex = None
+
+    try:
+        # get the title from the message, if it exists
+        # rare circumstance, but bibtex may not have a title
+        fallback_data = copy.copy(CITATION_FALLBACK_DATA)
+        if title := (
+            None if not message.get("title") else message.get("title", [None])[0]
+        ):
+            fallback_data["title"] = title
+
+        # TODO: we keep this for robustness, but likely not needed anymore,
+        # since we now create the bibtex from scratch
+        if query_bibtex:
+            bibtex = await doi_to_bibtex(
+                message["DOI"], session, missing_replacements=fallback_data  # type: ignore[arg-type]
+            )
+            # track the origin of the bibtex entry for debugging
+            bibtex_source = "crossref"
+
+    except DOINotFoundError:
+        pass
+
+    authors = [
+        f"{author.get('given', '')} {author.get('family', '')}".strip()
+        for author in message.get("author", [])
+    ]
+
+    publication_date = None
+    if "published" in message and "date-parts" in message["published"]:
+        date_parts = message["published"]["date-parts"][0]
+        if len(date_parts) >= 3:  # noqa: PLR2004
+            publication_date = datetime(date_parts[0], date_parts[1], date_parts[2])
+        elif len(date_parts) == 2:  # noqa: PLR2004
+            publication_date = datetime(date_parts[0], date_parts[1], 1)
+        elif len(date_parts) == 1:
+            publication_date = datetime(date_parts[0], 1, 1)
+
+    doc_details = DocDetails(  # type: ignore[call-arg]
```

Comment:
What do you think of making a `classmethod` of `DocDetails` for this?

Reply:
For this (and the semantic_scholar one) I was kind of thinking that it's the `MetadataProvider`'s job to deal with parsing. This means that to add a new provider, you just make the class and it gives you back a `DocDetails` object, with all of its logic encapsulated. If we move this logic into a `DocDetails` method, I think it implies that it could be accessed outside of the `clients` module which this re-factor is trying to remove as a possibility. LMK what you think on that.

Reply:
Okay I gotchu and sounds good, thanks!